accepted/0000-workspaces.md

```diff
@@ -0,0 +1,481 @@
+
+## Implementation
+
+### 1. Make npm workspace-aware
+
+We're following the lead of Yarn in supporting the `workspaces` `package.json` field which defines a list of paths, each of these paths may be a workspace itself but it also support globs.
+
+`package.json` example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": {
+        "packages": [
+            "packages/*"
+        ]
+    }
+}
+```
+
+`package.json` shorthand example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": [
+        "packages/*"
+    ]
+}
+```
+
+
+### 2. Installing dependencies across child packages
+
+Change `npm install` ([arborist](https://www.npmjs.com/package/@npmcli/arborist)) behavior to make it install every nested package by default once a valid configuration is defined.
+
+Arborist should also be aware of all nested workspaces and correctly link to an internal workspace should it match the required semver version of an expected dependency anywhere in the installing tree. e.g:
+
+```
+// Given this package.json structure:
+├── package.json { "workspaces": ["dep-a", "dep-b"] }
+├── dep-a
+│   └── package.json { "dependencies": { "dep-b": "^1.0.0" } }
+└── dep-b
+    └── package.json { "version": "1.3.1" }
+
+$ npm install
+
+// Results in this symlinking structure:
+├── node_modules
+│   ├── dep-a -> ./dep-a
+│   └── dep-b -> ./dep-b
+├── dep-a
+│   └── node_modules
+│       └── dep-b -> ../../../node_modules/dep-b
+└── dep-b
+```
+
+NOTE: The final implementation of the symlinking structure might differ from the examples laid out here (which are trying to be as illustrative as possible). Arborist might end up resolving all symlinks up to the actual place instead of creating a chain of symlinks.
+
+### 3. Run commands across all child packages
+
+Make npm commands workspace-aware, so that running a command tries to run it within nested packages as long as a workspaces field is defined in `package.json`.
+
+Only a subset of npm commands are to be supported:
+
+- `fund` List funding info for all packages
+- `ls` List all packages
+- `outdated` List outdated deps for all packages
+- `publish` Publishes all packages
+- `run-script` Run arbitrary scripts for all packages
+- `rebuild` Rebuild all packages
+- `restart`
+- `start`
+- `stop`
+- `test` Run tests in all packages
+- `update` Update a dependency in all packages
+- `version` Bump versions for all packages
+- `view` View registry info for all packages
+
+A new config value should be introduced in order to allow for filtering out a subset of the packages in which to run these commands. e.g: `--filter`
```

Comment:
Is this flag for listing subdirectories that are excluded or subdirs that are included in the command?

Reply:
this is the blurriest part of the current proposal and that's on purpose since I couldn't come up with a good enough API design, ideally I would like for it to be very simplistic and intuitive - that said, my hope is that we'll get there during the time of this RFC 😊 

some prior art in this specific domain:
- [lerna filter-options](https://www.npmjs.com/package/@lerna/filter-options)
- [yarn v2 foreach include/exclude](https://yarnpkg.com/cli/workspaces/foreach)
- [pnpm recursive --filter](https://pnpm.js.org/en/cli/recursive#filter-lt-package_selector)

Reply:
@danielleadams let's move the discussion to the follow up PR #117 😊 

it elaborates a bit more on filtering **workspaces** and propose some solutions such as introducing a way to group them